File: blizzard-wasm-spec/reference/Arima.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Blizzard;

/// <summary>
/// Easter date calculation using the Anonymous Gregorian algorithm (Computus)
/// </summary>
public static class EasterCalculator
{
    /// <summary>
    /// Calculate Easter Sunday for a given year
    /// </summary>
    public static DateTime GetEasterSunday(int year)
    {
        // Anonymous Gregorian algorithm
        int a = year % 19;
        int b = year / 100;
        int c = year % 100;
        int d = b / 4;
        int e = b % 4;
        int f = (b + 8) / 25;
        int g = (b - f + 1) / 3;
        int h = (19 * a + b - d - g + 15) % 30;
        int i = c / 4;
        int k = c % 4;
        int l = (32 + 2 * e + 2 * i - h - k) % 7;
        int m = (a + 11 * h + 22 * l) / 451;
        int month = (h + l - 7 * m + 114) / 31;
        int day = ((h + l - 7 * m + 114) % 31) + 1;

        return new DateTime(year, month, day);
    }

    /// <summary>
    /// Get the invoice month for Easter (3 months before)
    /// </summary>
    public static (int year, int month) GetEasterInvoiceMonth(int easterYear)
    {
        var easter = GetEasterSunday(easterYear);
        var invoiceDate = easter.AddMonths(-3);
        return (invoiceDate.Year, invoiceDate.Month);
    }

    /// <summary>
    /// Create Easter regressor array for a time series
    /// Returns 1.0 for months that are Easter invoice months, 0.0 otherwise
    /// </summary>
    public static double[] CreateEasterRegressor(int startYear, int startMonth, int length)
    {
        var regressor = new double[length];

        // Pre-calculate Easter invoice months for relevant years
        var easterInvoiceMonths = new HashSet<(int year, int month)>();
        for (int year = startYear; year <= startYear + (length / 12) + 2; year++)
        {
            easterInvoiceMonths.Add(GetEasterInvoiceMonth(year));
        }

        // Fill regressor array
        int currentYear = startYear;
        int currentMonth = startMonth;

        for (int i = 0; i < length; i++)
        {
            regressor[i] = easterInvoiceMonths.Contains((currentYear, currentMonth)) ? 1.0 : 0.0;

            currentMonth++;
            if (currentMonth > 12)
            {
                currentMonth = 1;
                currentYear++;
            }
        }

        return regressor;
    }
}

/// <summary>
/// ARIMA(p,d,q) model implementation with seasonal support and optional exogenous variables (ARIMAX)
/// </summary>
public class Arima
{
    public int P { get; } // AR order
    public int D { get; } // Differencing order
    public int Q { get; } // MA order
    public int SeasonalPeriod { get; } // Seasonal period (12 for monthly)

    private double[] _arCoeffs = Array.Empty<double>();
    private double[] _maCoeffs = Array.Empty<double>();
    private double[] _seasonalFactors = Array.Empty<double>();
    private double _intercept;
    private double[] _originalSeries = Array.Empty<double>();
    private double[] _differencedSeries = Array.Empty<double>();
    private double[] _residuals = Array.Empty<double>();

    // ARIMAX support
    private double[] _exogCoeffs = Array.Empty<double>();
    private double[]? _exogData;
    private int _exogCount;

    /// <summary>
    /// Exogenous variable coefficients (for ARIMAX)
    /// </summary>
    public double[] ExogCoefficients => _exogCoeffs;

    public Arima(int p = 2, int d = 1, int q = 1, int seasonalPeriod = 12)
    {
        P = p;
        D = d;
        Q = q;
        SeasonalPeriod = seasonalPeriod;
    }

    /// <summary>
    /// Fit ARIMA model without exogenous variables
    /// </summary>
    public void Fit(double[] series)
    {
        Fit(series, null);
    }

    /// <summary>
    /// Fit ARIMAX model with optional exogenous variables
    /// </summary>
    /// <param name="series">The time series data</param>
    /// <param name="exog">Optional exogenous variables (one array per variable, stacked)</param>
    public void Fit(double[] series, double[][]? exog)
    {
        if (series.Length < P + D + Q + SeasonalPeriod)
            throw new ArgumentException("Series too short for specified ARIMA parameters");

        _originalSeries = series.ToArray();
        _exogCount = exog?.Length ?? 0;
        _exogCoeffs = new double[_exogCount];

        // Store exog data for later use
        if (exog != null && exog.Length > 0)
        {
            _exogData = exog[0].ToArray();
        }

        // IMPORTANT: Regress out exogenous effects FIRST (before deseasonalizing)
        // The exog effect (e.g., Easter spike) is additive on the original scale
        double[] adjustedSeries = series.ToArray();
        if (exog != null && exog.Length > 0)
        {
            adjustedSeries = RegressOutExogenous(adjustedSeries, exog);
        }

        // Calculate seasonal factors on exog-adjusted series
        _seasonalFactors = CalculateSeasonalFactors(adjustedSeries);

        // Deseasonalize
        var deseasonalized = Deseasonalize(adjustedSeries, _seasonalFactors);

        // Apply differencing to deseasonalized series
        _differencedSeries = Difference(deseasonalized, D);

        // Estimate AR and MA coefficients using Yule-Walker and residuals
        EstimateCoefficients(_differencedSeries);
    }

    /// <summary>
    /// Regress out exogenous variables using simple mean-difference approach
    /// For sparse binary variables (like Easter), this is more accurate than
    /// standard OLS within ARIMAX which gets confused by differencing.
    ///
    /// Coefficient = mean(Y when X=1) - mean(Y when X=0)
    /// </summary>
    private double[] RegressOutExogenous(double[] series, double[][] exog)
    {
        int n = series.Length;
        int k = exog.Length;

        var residuals = series.ToArray();

        for (int j = 0; j < k; j++)
        {
            var x = exog[j];
            if (x.Length != n)
                throw new ArgumentException($"Exogenous variable {j} length mismatch");

            // For sparse binary exogenous variables, use mean difference
            // This is equivalent to OLS for binary X but more numerically stable
            var withExog = new List<double>();
            var withoutExog = new List<double>();

            for (int i = 0; i < n; i++)
            {
                if (x[i] > 0.5)
                    withExog.Add(residuals[i]);
                else
                    withoutExog.Add(residuals[i]);
            }

            if (withExog.Count > 0 && withoutExog.Count > 0)
            {
                double meanWith = withExog.Average();
                double meanWithout = withoutExog.Average();
                _exogCoeffs[j] = meanWith - meanWithout;

                // Remove exogenous effect from series (only from affected points)
                for (int i = 0; i < n; i++)
                {
                    if (x[i] > 0.5)
                    {
                        residuals[i] -= _exogCoeffs[j];
                    }
                }
            }
            else
            {
                _exogCoeffs[j] = 0;
            }
        }

        return residuals;
    }

    private double[] CalculateSeasonalFactors(double[] series)
    {
        var factors = new double[SeasonalPeriod];
        var counts = new int[SeasonalPeriod];
        var overall = series.Where(x => x > 0).Average();

        for (int i = 0; i < series.Length; i++)
        {
            if (series[i] > 0)
            {
                factors[i % SeasonalPeriod] += series[i];
                counts[i % SeasonalPeriod]++;
            }
        }

        for (int i = 0; i < SeasonalPeriod; i++)
        {
            factors[i] = counts[i] > 0 ? (factors[i] / counts[i]) / overall : 1.0;
        }

        return factors;
    }

    private double[] Deseasonalize(double[] series, double[] factors)
    {
        var result = new double[series.Length];
        for (int i = 0; i < series.Length; i++)
        {
            var factor = factors[i % SeasonalPeriod];
            result[i] = factor > 0 ? series[i] / factor : series[i];
        }
        return result;
    }

    private double[] Reseasonalize(double[] series, double[] factors, int startIndex)
    {
        var result = new double[series.Length];
        for (int i = 0; i < series.Length; i++)
        {
            result[i] = series[i] * factors[(startIndex + i) % SeasonalPeriod];
        }
        return result;
    }

    private double[] Difference(double[] series, int order)
    {
        var result = series.ToArray();
        for (int d = 0; d < order; d++)
        {
            var temp = new double[result.Length - 1];
            for (int i = 1; i < result.Length; i++)
            {
                temp[i - 1] = result[i] - result[i - 1];
            }
            result = temp;
        }
        return result;
    }

    private double[] Undifference(double[] diffSeries, double[] originalSeries, int order)
    {
        if (order == 0) return diffSeries;

        var result = diffSeries.ToArray();
        for (int d = 0; d < order; d++)
        {
            var temp = new double[result.Length];
            double lastValue = originalSeries[originalSeries.Length - 1 - (order - d - 1)];

            for (int i = 0; i < result.Length; i++)
            {
                temp[i] = result[i] + lastValue;
                lastValue = temp[i];
            }
            result = temp;
        }
        return result;
    }

    private void EstimateCoefficients(double[] series)
    {
        _intercept = series.Average();
        var centered = series.Select(x => x - _intercept).ToArray();

        // Yule-Walker for AR coefficients
        _arCoeffs = new double[P];
        if (P > 0)
        {
            var autocorr = CalculateAutocorrelation(centered, P);
            _arCoeffs = SolveYuleWalker(autocorr);
        }

        // Calculate residuals for MA estimation
        _residuals = CalculateResiduals(centered, _arCoeffs);

        // Estimate MA coefficients from residual autocorrelation
        _maCoeffs = new double[Q];
        if (Q > 0)
        {
            var resAutocorr = CalculateAutocorrelation(_residuals, Q);
            for (int i = 0; i < Q; i++)
            {
                _maCoeffs[i] = resAutocorr[i + 1] * 0.5; // Simplified MA estimation
            }
        }
    }

    private double[] CalculateAutocorrelation(double[] series, int maxLag)
    {
        var result = new double[maxLag + 1];
        double variance = series.Select(x => x * x).Average();

        if (variance < 1e-10)
        {
            result[0] = 1.0;
            return result;
        }

        for (int lag = 0; lag <= maxLag; lag++)
        {
            double sum = 0;
            for (int i = lag; i < series.Length; i++)
            {
                sum += series[i] * series[i - lag];
            }
            result[lag] = sum / (series.Length * variance);
        }

        return result;
    }

    private double[] SolveYuleWalker(double[] autocorr)
    {
        // Levinson-Durbin algorithm
        int p = autocorr.Length - 1;
        if (p == 0) return Array.Empty<double>();

        var phi = new double[p];
        var phiPrev = new double[p];

        phi[0] = autocorr[1];
        double v = 1 - phi[0] * phi[0];

        for (int i = 1; i < p; i++)
        {
            Array.Copy(phi, phiPrev, p);

            double num = autocorr[i + 1];
            for (int j = 0; j < i; j++)
            {
                num -= phiPrev[j] * autocorr[i - j];
            }

            phi[i] = num / v;

            for (int j = 0; j < i; j++)
            {
                phi[j] = phiPrev[j] - phi[i] * phiPrev[i - 1 - j];
            }

            v *= (1 - phi[i] * phi[i]);
            if (v < 1e-10) break;
        }

        return phi;
    }

    private double[] CalculateResiduals(double[] series, double[] arCoeffs)
    {
        var residuals = new double[series.Length];

        for (int i = 0; i < series.Length; i++)
        {
            double predicted = 0;
            for (int j = 0; j < arCoeffs.Length && i - j - 1 >= 0; j++)
            {
                predicted += arCoeffs[j] * series[i - j - 1];
            }
            residuals[i] = series[i] - predicted;
        }

        return residuals;
    }

    /// <summary>
    /// Forecast without exogenous variables
    /// </summary>
    public double[] Forecast(int steps)
    {
        return Forecast(steps, null);
    }

    /// <summary>
    /// Forecast with optional exogenous variables for future periods
    /// </summary>
    /// <param name="steps">Number of periods to forecast</param>
    /// <param name="futureExog">Exogenous values for forecast periods (one array per variable)</param>
    public double[] Forecast(int steps, double[][]? futureExog)
    {
        // Work with deseasonalized differenced series
        var extended = _differencedSeries.ToList();
        var extendedResiduals = _residuals.ToList();

        for (int s = 0; s < steps; s++)
        {
            double prediction = _intercept;

            // AR component
            for (int i = 0; i < P && i < extended.Count; i++)
            {
                prediction += _arCoeffs[i] * (extended[extended.Count - 1 - i] - _intercept);
            }

            // MA component (residuals decay to 0 for forecasts)
            for (int i = 0; i < Q && i < extendedResiduals.Count; i++)
            {
                if (extended.Count - 1 - i < _differencedSeries.Length)
                {
                    prediction += _maCoeffs[i] * extendedResiduals[extendedResiduals.Count - 1 - i];
                }
            }

            extended.Add(prediction);
            extendedResiduals.Add(0); // Future residuals are 0
        }

        // Extract forecast values
        var forecastDiff = extended.Skip(_differencedSeries.Length).ToArray();

        // Undifference
        var deseasonalized = _originalSeries.Select((v, i) =>
            _seasonalFactors[i % SeasonalPeriod] > 0 ? v / _seasonalFactors[i % SeasonalPeriod] : v).ToArray();
        var forecastDeseas = Undifference(forecastDiff, deseasonalized, D);

        // Reseasonalize
        int startMonth = _originalSeries.Length % SeasonalPeriod;
        var forecast = Reseasonalize(forecastDeseas, _seasonalFactors, startMonth);

        // Add back exogenous effects for future periods
        if (futureExog != null && _exogCoeffs.Length > 0)
        {
            for (int s = 0; s < steps && s < forecast.Length; s++)
            {
                for (int j = 0; j < _exogCoeffs.Length && j < futureExog.Length; j++)
                {
                    if (s < futureExog[j].Length)
                    {
                        // Add exogenous effect (coefficient * future value)
                        forecast[s] += _exogCoeffs[j] * futureExog[j][s];
                    }
                }
            }
        }

        // Ensure non-negative
        return forecast.Select(x => Math.Max(0, x)).ToArray();
    }

    public (double[] Lower, double[] Upper) ForecastConfidenceInterval(int steps, double confidence = 0.95)
    {
        var forecast = Forecast(steps);

        // Estimate standard error from residuals
        double se = Math.Sqrt(_residuals.Select(r => r * r).Average());

        // Z-score for confidence level
        double z = confidence switch
        {
            0.99 => 2.576,
            0.95 => 1.96,
            0.90 => 1.645,
            0.80 => 1.28,
            _ => 1.96
        };

        var lower = new double[steps];
        var upper = new double[steps];

        for (int i = 0; i < steps; i++)
        {
            // Error grows with forecast horizon
            double horizon_se = se * Math.Sqrt(1 + i * 0.1);
            // Scale by seasonal factor for proper interval width
            double seasonalScale = _seasonalFactors[(_originalSeries.Length + i) % SeasonalPeriod];
            double interval = z * horizon_se * seasonalScale;

            lower[i] = Math.Max(0, forecast[i] - interval);
            upper[i] = forecast[i] + interval;
        }

        return (lower, upper);
    }

    public ForecastMetrics CalculateMetrics(double[] actual, double[] predicted)
    {
        int n = Math.Min(actual.Length, predicted.Length);

        double mse = 0, mae = 0, mape = 0;
        int mapeCount = 0;

        for (int i = 0; i < n; i++)
        {
            double error = actual[i] - predicted[i];
            mse += error * error;
            mae += Math.Abs(error);
            if (actual[i] != 0)
            {
                mape += Math.Abs(error / actual[i]);
                mapeCount++;
            }
        }

        return new ForecastMetrics
        {
            MSE = mse / n,
            RMSE = Math.Sqrt(mse / n),
            MAE = mae / n,
            MAPE = mapeCount > 0 ? (mape / mapeCount) * 100 : 0
        };
    }
}

public class ForecastMetrics
{
    public double MSE { get; set; }
    public double RMSE { get; set; }
    public double MAE { get; set; }
    public double MAPE { get; set; }
}

File: blizzard-wasm/src/arima.rs
//! ARIMA(p,d,q) model implementation with seasonal support and ARIMAX extensions
//!
//! This module ports the C# Arima.cs implementation to Rust for WASM compilation.
//! The model uses:
//! - p=2 AR terms
//! - d=1 differencing
//! - q=1 MA term
//! - Seasonal period of 12 (monthly data)
//! - Optional Easter regressor for ARIMAX

use crate::easter::create_easter_regressor;

/// Result of fitting and forecasting with ARIMA
#[derive(Debug, Clone)]
pub struct ForecastResult {
    /// Point forecasts for each future period
    pub forecast: Vec<f64>,
    /// Lower bound of 80% confidence interval
    pub lower: Vec<f64>,
    /// Upper bound of 80% confidence interval
    pub upper: Vec<f64>,
    /// Seasonal factors (12 values for monthly data)
    pub seasonal_factors: Vec<f64>,
    /// Estimated Easter effect coefficient (if ARIMAX)
    pub easter_coefficient: f64,
    /// Estimated AR coefficients
    pub ar_coefficients: Vec<f64>,
    /// Estimated MA coefficients
    pub ma_coefficients: Vec<f64>,
    /// Model intercept
    pub intercept: f64,
}

/// ARIMA model with optional exogenous variables
pub struct Arima {
    p: usize,              // AR order
    d: usize,              // Differencing order
    q: usize,              // MA order
    seasonal_period: usize, // Seasonal period (12 for monthly)
    
    // Fitted values (populated after fit())
    ar_coeffs: Vec<f64>,
    ma_coeffs: Vec<f64>,
    seasonal_factors: Vec<f64>,
    intercept: f64,
    original_series: Vec<f64>,
    differenced_series: Vec<f64>,
    residuals: Vec<f64>,
    
    // ARIMAX support
    exog_coeffs: Vec<f64>,
    exog_data: Option<Vec<f64>>,
}

impl Arima {
    /// Create a new ARIMA model with specified parameters
    pub fn new(p: usize, d: usize, q: usize, seasonal_period: usize) -> Self {
        Arima {
            p,
            d,
            q,
            seasonal_period,
            ar_coeffs: vec![],
            ma_coeffs: vec![],
            seasonal_factors: vec![],
            intercept: 0.0,
            original_series: vec![],
            differenced_series: vec![],
            residuals: vec![],
            exog_coeffs: vec![],
            exog_data: None,
        }
    }

    /// Fit the model to a time series
    pub fn fit(&mut self, series: &[f64]) {
        self.fit_with_exog(series, None);
    }

    /// Fit the model with optional exogenous variables
    pub fn fit_with_exog(&mut self, series: &[f64], exog: Option<&[f64]>) {
        self.original_series = series.to_vec();
        self.exog_data = exog.map(|e| e.to_vec());

        // 1. Regress out exogenous effects (if present)
        let adjusted_series = if let Some(exog_data) = exog {
            let (adj, coef) = regress_out_exogenous(series, exog_data);
            self.exog_coeffs = vec![coef];
            adj
        } else {
            self.exog_coeffs = vec![];
            series.to_vec()
        };

        // 2. Calculate seasonal factors
        self.seasonal_factors = calculate_seasonal_factors(&adjusted_series, self.seasonal_period);

        // 3. Deseasonalize
        let deseasonalized = deseasonalize(&adjusted_series, &self.seasonal_factors);

        // 4. Apply differencing
        self.differenced_series = difference(&deseasonalized, self.d);

        // 5. Estimate AR/MA coefficients
        self.intercept = mean(&self.differenced_series);
        let centered: Vec<f64> = self.differenced_series.iter()
            .map(|&x| x - self.intercept)
            .collect();

        // Yule-Walker for AR coefficients
        if self.p > 0 {
            let autocorr = autocorrelation(&centered, self.p);
            self.ar_coeffs = solve_yule_walker(&autocorr);
        } else {
            self.ar_coeffs = vec![];
        }

        // Calculate residuals for MA estimation
        self.residuals = calculate_residuals(&centered, &self.ar_coeffs);

        // Estimate MA coefficients from residual autocorrelation
        if self.q > 0 {
            self.ma_coeffs = estimate_ma_coefficients(&self.residuals, self.q);
        } else {
            self.ma_coeffs = vec![];
        }
    }

    /// Generate forecasts
    pub fn forecast(&self, steps: usize) -> Vec<f64> {
        self.forecast_with_exog(steps, None)
    }

    /// Generate forecasts with future exogenous values
    pub fn forecast_with_exog(&self, steps: usize, future_exog: Option<&[f64]>) -> Vec<f64> {
        // Work with differenced series
        let mut extended = self.differenced_series.clone();
        let mut extended_residuals = self.residuals.clone();

        // 1. Forecast differenced series
        for _ in 0..steps {
            let mut prediction = self.intercept;

            // AR component
            for (i, &coef) in self.ar_coeffs.iter().enumerate() {
                if i < extended.len() {
                    prediction += coef * (extended[extended.len() - 1 - i] - self.intercept);
                }
            }

            // MA component (residuals decay to 0 for forecasts)
            for (i, &coef) in self.ma_coeffs.iter().enumerate() {
                let res_idx = extended_residuals.len() - 1 - i;
                if res_idx < self.residuals.len() {
                    prediction += coef * extended_residuals[res_idx];
                }
            }

            extended.push(prediction);
            extended_residuals.push(0.0); // Future residuals are 0
        }

        // Extract forecast values
        let forecast_diff: Vec<f64> = extended[self.differenced_series.len()..].to_vec();

        // 2. Undifference - get deseasonalized original series first
        let deseasonalized: Vec<f64> = self.original_series.iter().enumerate()
            .map(|(i, &v)| {
                let factor = self.seasonal_factors[i % self.seasonal_period];
                if factor > 0.0 { v / factor } else { v }
            })
            .collect();
        let forecast_deseas = undifference(&forecast_diff, &deseasonalized, self.d);

        // 3. Reseasonalize
        let start_month = self.original_series.len() % self.seasonal_period;
        let mut forecast = reseasonalize(&forecast_deseas, &self.seasonal_factors, start_month);

        // 4. Add back exogenous effects for future periods
        if let Some(future_exog_data) = future_exog {
            if !self.exog_coeffs.is_empty() {
                for (s, f) in forecast.iter_mut().enumerate() {
                    if s < future_exog_data.len() {
                        *f += self.exog_coeffs[0] * future_exog_data[s];
                    }
                }
            }
        }

        // Ensure non-negative
        forecast.iter().map(|&x| x.max(0.0)).collect()
    }

    /// Calculate confidence intervals for forecasts
    pub fn confidence_intervals(&self, steps: usize, confidence: f64) -> (Vec<f64>, Vec<f64>) {
        let forecast = self.forecast(steps);

        // Estimate standard error from residuals
        let se = (self.residuals.iter().map(|&r| r * r).sum::<f64>() / self.residuals.len().max(1) as f64).sqrt();

        // Z-score for confidence level
        let z = if (confidence - 0.99).abs() < 0.001 {
            2.576
        } else if (confidence - 0.95).abs() < 0.001 {
            1.96
        } else if (confidence - 0.90).abs() < 0.001 {
            1.645
        } else if (confidence - 0.80).abs() < 0.001 {
            1.28
        } else {
            1.96
        };

        let lower: Vec<f64> = forecast.iter().enumerate().map(|(i, &f)| {
            // Error grows with forecast horizon
            let horizon_se = se * (1.0 + i as f64 * 0.1).sqrt();
            // Scale by seasonal factor for proper interval width
            let seasonal_scale = self.seasonal_factors[(self.original_series.len() + i) % self.seasonal_period];
            let interval = z * horizon_se * seasonal_scale;
            (f - interval).max(0.0)
        }).collect();

        let upper: Vec<f64> = forecast.iter().enumerate().map(|(i, &f)| {
            let horizon_se = se * (1.0 + i as f64 * 0.1).sqrt();
            let seasonal_scale = self.seasonal_factors[(self.original_series.len() + i) % self.seasonal_period];
            let interval = z * horizon_se * seasonal_scale;
            f + interval
        }).collect();

        (lower, upper)
    }

    /// Get the estimated exogenous coefficients
    pub fn exog_coefficients(&self) -> &[f64] {
        &self.exog_coeffs
    }
}

// ============================================================================
// Helper functions - implement these first, they're used by Arima
// ============================================================================

/// Calculate mean of a slice
fn mean(data: &[f64]) -> f64 {
    if data.is_empty() {
        return 0.0;
    }
    data.iter().sum::<f64>() / data.len() as f64
}

/// Regress out exogenous variables using mean-difference approach
///
/// For sparse binary exogenous variables (like Easter), this is more stable
/// than standard OLS within ARIMAX.
///
/// Returns (adjusted_series, coefficient)
fn regress_out_exogenous(series: &[f64], exog: &[f64]) -> (Vec<f64>, f64) {
    let mut residuals = series.to_vec();

    // Separate observations by exog value
    let with_exog: Vec<f64> = series.iter().zip(exog.iter())
        .filter(|(_, &x)| x > 0.5)
        .map(|(&y, _)| y)
        .collect();

    let without_exog: Vec<f64> = series.iter().zip(exog.iter())
        .filter(|(_, &x)| x <= 0.5)
        .map(|(&y, _)| y)
        .collect();

    let coefficient = if !with_exog.is_empty() && !without_exog.is_empty() {
        mean(&with_exog) - mean(&without_exog)
    } else {
        0.0
    };

    // Remove effect from affected observations
    for (i, &x) in exog.iter().enumerate() {
        if x > 0.5 {
            residuals[i] -= coefficient;
        }
    }

    (residuals, coefficient)
}

/// Calculate multiplicative seasonal factors
fn calculate_seasonal_factors(series: &[f64], period: usize) -> Vec<f64> {
    let mut sums = vec![0.0; period];
    let mut counts = vec![0usize; period];

    // Only include positive values
    let overall_mean = series.iter().filter(|&&x| x > 0.0).sum::<f64>()
        / series.iter().filter(|&&x| x > 0.0).count().max(1) as f64;

    for (i, &value) in series.iter().enumerate() {
        if value > 0.0 {
            let month_idx = i % period;
            sums[month_idx] += value;
            counts[month_idx] += 1;
        }
    }

    // Factor = month_average / overall_average
    (0..period).map(|i| {
        if counts[i] > 0 {
            (sums[i] / counts[i] as f64) / overall_mean
        } else {
            1.0
        }
    }).collect()
}

/// Deseasonalize series by dividing by seasonal factors
fn deseasonalize(series: &[f64], factors: &[f64]) -> Vec<f64> {
    series.iter().enumerate().map(|(i, &value)| {
        let factor = factors[i % factors.len()];
        if factor > 0.0 { value / factor } else { value }
    }).collect()
}

/// Reseasonalize series by multiplying by seasonal factors
fn reseasonalize(series: &[f64], factors: &[f64], start_idx: usize) -> Vec<f64> {
    series.iter().enumerate().map(|(i, &value)| {
        let factor_idx = (start_idx + i) % factors.len();
        value * factors[factor_idx]
    }).collect()
}

/// Apply d-order differencing
fn difference(series: &[f64], d: usize) -> Vec<f64> {
    let mut result = series.to_vec();
    
    for _ in 0..d {
        let temp: Vec<f64> = (1..result.len())
            .map(|i| result[i] - result[i - 1])
            .collect();
        result = temp;
    }
    
    result
}

/// Undo differencing to get back to original scale
fn undifference(differenced: &[f64], original: &[f64], d: usize) -> Vec<f64> {
    if d == 0 {
        return differenced.to_vec();
    }

    let mut result = differenced.to_vec();

    for ord in 0..d {
        let last_value = original[original.len() - 1 - (d - ord - 1)];
        let mut cumsum = vec![last_value + result[0]];

        for i in 1..result.len() {
            cumsum.push(cumsum[i - 1] + result[i]);
        }
        result = cumsum;
    }

    result
}

/// Calculate autocorrelation function up to max_lag
fn autocorrelation(series: &[f64], max_lag: usize) -> Vec<f64> {
    let n = series.len();
    let variance = series.iter().map(|&x| x * x).sum::<f64>() / n as f64;

    if variance < 1e-10 {
        let mut result = vec![0.0; max_lag + 1];
        result[0] = 1.0;
        return result;
    }

    (0..=max_lag).map(|lag| {
        let sum: f64 = (lag..n)
            .map(|i| series[i] * series[i - lag])
            .sum();
        sum / (n as f64 * variance)
    }).collect()
}

/// Solve Yule-Walker equations using Levinson-Durbin algorithm
fn solve_yule_walker(autocorr: &[f64]) -> Vec<f64> {
    let p = autocorr.len() - 1;
    if p == 0 {
        return vec![];
    }

    let mut phi = vec![0.0; p];
    let mut phi_prev = vec![0.0; p];

    // Initialize
    phi[0] = autocorr[1];
    let mut v = 1.0 - phi[0] * phi[0];

    for i in 1..p {
        phi_prev.copy_from_slice(&phi);

        // Calculate reflection coefficient
        let mut num = autocorr[i + 1];
        for j in 0..i {
            num -= phi_prev[j] * autocorr[i - j];
        }

        phi[i] = num / v;

        // Update coefficients
        for j in 0..i {
            phi[j] = phi_prev[j] - phi[i] * phi_prev[i - 1 - j];
        }

        v *= 1.0 - phi[i] * phi[i];
        if v < 1e-10 {
            break;
        }
    }

    phi
}

/// Calculate residuals from AR model
fn calculate_residuals(series: &[f64], ar_coeffs: &[f64]) -> Vec<f64> {
    series.iter().enumerate().map(|(i, &value)| {
        let predicted: f64 = ar_coeffs.iter().enumerate()
            .filter(|(j, _)| i > *j)
            .map(|(j, &coef)| coef * series[i - j - 1])
            .sum();
        value - predicted
    }).collect()
}

/// Estimate MA coefficients from residual autocorrelation
fn estimate_ma_coefficients(residuals: &[f64], q: usize) -> Vec<f64> {
    if q == 0 {
        return vec![];
    }

    let autocorr = autocorrelation(residuals, q);

    // Simplified: MA coefficient ≈ 0.5 × residual autocorrelation
    (1..=q).map(|i| autocorr[i] * 0.5).collect()
}

// ============================================================================
// High-level convenience function
// ============================================================================

/// Fit ARIMA model and generate forecast in one call
///
/// This is the main entry point for the WASM interface.
pub fn fit_and_forecast(
    series: &[f64],
    start_year: i32,
    start_month: u32,
    forecast_months: usize,
    use_easter: bool,
) -> ForecastResult {
    let mut model = Arima::new(2, 1, 1, 12);
    
    let (easter_coef, adjusted_series) = if use_easter {
        let regressor = create_easter_regressor(start_year, start_month, series.len());
        let (adj, coef) = regress_out_exogenous(series, &regressor);
        (coef, adj)
    } else {
        (0.0, series.to_vec())
    };
    
    model.fit(&adjusted_series);
    
    // Generate future Easter regressor if needed
    let forecast = if use_easter {
        let last_idx = series.len();
        let last_month = ((start_month as usize - 1 + last_idx) % 12 + 1) as u32;
        let years_elapsed = (start_month as usize - 1 + last_idx) / 12;
        let last_year = start_year + years_elapsed as i32;
        
        let future_easter = create_easter_regressor(last_year, last_month + 1, forecast_months);
        model.forecast_with_exog(forecast_months, Some(&future_easter))
    } else {
        model.forecast(forecast_months)
    };
    
    let (lower, upper) = model.confidence_intervals(forecast_months, 0.80);
    
    ForecastResult {
        forecast,
        lower,
        upper,
        seasonal_factors: model.seasonal_factors.clone(),
        easter_coefficient: easter_coef,
        ar_coefficients: model.ar_coeffs.clone(),
        ma_coefficients: model.ma_coeffs.clone(),
        intercept: model.intercept,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mean() {
        assert!((mean(&[1.0, 2.0, 3.0, 4.0, 5.0]) - 3.0).abs() < 1e-10);
        assert!(mean(&[]).abs() < 1e-10);
    }

    #[test]
    fn test_deseasonalize() {
        let series = vec![100.0, 120.0, 90.0, 110.0];
        let factors = vec![1.0, 1.2, 0.9, 1.1];
        let result = deseasonalize(&series, &factors);
        
        // Each value should become 100.0 after deseasonalization
        for &val in &result {
            assert!((val - 100.0).abs() < 1e-10);
        }
    }

    #[test]
    fn test_difference() {
        let series = vec![10.0, 12.0, 15.0, 14.0, 18.0];
        let diff = difference(&series, 1);
        
        // First differences: [2, 3, -1, 4]
        assert_eq!(diff.len(), 4);
        assert!((diff[0] - 2.0).abs() < 1e-10);
        assert!((diff[1] - 3.0).abs() < 1e-10);
        assert!((diff[2] - (-1.0)).abs() < 1e-10);
        assert!((diff[3] - 4.0).abs() < 1e-10);
    }

    // TODO: Add more tests as functions are implemented
}
